apps/application-system/api/src/app/modules/application/lifecycle/application-lifecycle.service.ts

```diff
@@ -0,0 +1,165 @@
+import { Inject, Injectable } from '@nestjs/common'
+import { APPLICATION_CONFIG } from '../application.configuration'
+import type { ApplicationConfig } from '../application.configuration'
+import { Application } from '../application.model'
+import { ApplicationService } from '../application.service'
+import { AwsService } from '../files'
+import AmazonS3URI from 'amazon-s3-uri'
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import type { Logger } from '@island.is/logging'
+
+export interface ApplicationPruning {
+  pruned: boolean
+  application: Application
+  failedAttachments: object
+}
+
+@Injectable()
+export class ApplicationLifeCycleService {
+  private processingApplications: ApplicationPruning[] = []
+
+  constructor(
+    @Inject(APPLICATION_CONFIG)
+    private readonly config: ApplicationConfig,
+    @Inject(LOGGER_PROVIDER)
+    private logger: Logger,
+    private applicationService: ApplicationService,
+    private awsService: AwsService,
+  ) {}
+
+  public async run() {
+    this.logger.info(`Starting application pruning...`)
+    await this.fetchApplicationsToBePruned()
+    await this.pruneAttachments()
+    await this.pruneApplicationData()
+    this.reportResults()
+    this.logger.info(`Application pruning done.`)
+  }
+
+  public getProcessingApplications() {
+    return this.processingApplications
+  }
+
+  private async getJustOneById(id: string) {
+    const application = await this.applicationService.findOneById(id)
+    if (application) {
+      this.processingApplications.push({
+        pruned: false,
+        application,
+        failedAttachments: {},
+      })
+    }
+  }
+
+  private async fetchApplicationsToBePruned() {
+    const applications = await this.applicationService.findAllDueToBePruned()
+    this.logger.info(`Found ${applications.length} applications to be pruned.`)
+    this.processingApplications = applications.map((application) => {
+      return {
+        pruned: true,
+        application,
+        failedAttachments: {},
+      }
+    })
+  }
+
+  private async pruneAttachments() {
+    for (const prune of this.processingApplications) {
+      const applicationAttachments = prune.application.attachments as {
+        key: string
+        name: string
+      }
+
+      const attachments = this.attachmentsToKeyArray(applicationAttachments)
+      try {
+        if (attachments) {
+          await this.awsService.deleteObjects(this.getBucketName(), attachments)
+        }
```

Comment:
Maybe simplify this logic to use `deleteObject`, handle any error with `failedAttachments` and skip the extra `fileExists` check.

Then we can get the bucket name from each attachment's url using AmazonS3URI since some attachments may be stuck in the upload bucket which might fail the other command. Though I'm not sure if this service account has delete permissions for the upload bucket ...

Note that the upload bucket has its own lifecycle rules so the object might be already gone. Would be great to verify what deleteObject does for objects that are already deleted or don't exist for some other reason.

Reply:
Well the  the `deleteObjects` and `deleteObject` just returns success if the file does not exist. So it wont tell us if we actually removed it or it never was there to begin with.
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property

Reply:
Switched over to use `deleteObject` and  `AmazonS3URI` to determine the bucket name. But I kept the `fileExists` check because `deleteObject` does not appear to be giving us a good enough feedback on wether the file was delted or not. I still however worry that this will amount to alot of calls to the aws sdk for a single run.